src/neptune/internal/backends/hosted_file_operations.py

```diff
@@ -460,18 +467,29 @@ def _get_download_url(swagger_client: SwaggerClientWrapper, download_id: str):
     return download_request.downloadUrl
 
 
-def _store_response_as_file(response: Response, destination: Optional[str] = None):
+def _store_response_as_file(
+    response: Response,
+    destination: Optional[str] = None,
+    pre_download_hook: Callable[[int], None] = lambda x: None,
+    download_iter_hook: Callable[[int], None] = lambda x: None,
+    post_download_hook: Callable[[], None] = lambda: None,
+) -> None:
     if destination is None:
         target_file = _get_content_disposition_filename(response)
     elif os.path.isdir(destination):
         target_file = os.path.join(destination, _get_content_disposition_filename(response))
     else:
         target_file = destination
+
+    total_size = int(response.headers.get("content-length", 0))
+    pre_download_hook(total_size)
     with response:
         with open(target_file, "wb") as f:
             for chunk in response.iter_content(chunk_size=1024 * 1024):
                 if chunk:
                     f.write(chunk)
+                download_iter_hook(len(chunk))
```

Comment:
`len(chunk)` is risky outside of `if chunk:`. Use `len(chunk) if chunk else 0` or pass `chunk` to the hook directly. Not sure if `iter_content` could return `None` or something similar that will break `len(chunk)` with some `ValueError`

Reply:
Good point